---
title: "Harwell PRDM16 RNA-Seq: Tbr2+"
output:
  html_document:
    theme: cosmo
    toc: true
    toc_depth: 4
    fig_caption: true
    fig_width: 7
    fig_height: 6
author: "Meeta Mistry"
---

```{r setup, echo=FALSE}
 
# Setup report details
clientname="Manuel Baizabal"
clientemail="manuel_baizabal@hms.harvard.edu "
lablocation="Harwell lab" 
analystname="Meeta Mistry"
analystemail="mmistry@hsph.harvard.edu"
```

RNA-Seq analysis for `r clientname` (`r clientemail`) at `r lablocation`. Contact `r analystname` (`r analystemail`) for additional details. Request from client was:     

> Study is to purify committed neuronal progenitor cells using Tbr2 as marker and identify transcriptional changes associated with Pdrm16 and how this compares to other cortical neurons (Neg).  Experimental setup:
> 
> * 4 x WT Tbr2+
> * 4 x Prdm16 KO Tbr2+
> * 4 x WT Neg
> * 4 x Prdm16 KO Neg

## Setup
### Bioconductor and R libraries used

```{r libraries, echo=TRUE}
loadlibs <- function(){
library(ggplot2)
library(reshape)
library(gplots)
library(pheatmap)
library(RColorBrewer)
library(CHBUtils)
library(ggdendro)
library(grid)
library(gridExtra)
library(DESeq2)
library(Biobase)
library(genefilter)
library(CHBUtils)
library(gProfileR)
source('revigo.R') ## for GO analysis
}
suppressPackageStartupMessages(loadlibs())
```

### Set variables
```{r directories, echo=TRUE}
baseDir=getwd()
dataDir=paste(baseDir, "/data/", sep="")
resultsDir=paste(baseDir, "/results/", sep="")
metaDir=paste(baseDir, "/meta/", sep="")

heatcolors.1 <- rev(brewer.pal(6, "YlOrRd"))
cbPalette <- cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", 
                            "#D55E00", "#CC79A7", "#000000")

p.cutoff <- 0.05
fc.cutoff <- 1.5
```

### Load data

```{r loadcounts}
data <- read.table(file.path(resultsDir, 'tbr2/combined_runs.counts'), header=T)
meta <- read.table(file.path(resultsDir, 'tbr2/combined_runs.metadata'), header=T)

# Get unique samples
meta$samples <- factor(paste(meta$celltype, meta$genotype, meta$sort, sep="_"))
meta$sort <- factor(meta$sort)

```

### Aggregate data
Each sample was run over three lanes except for neg_WT_5 which has only data from two lanes (due to sample mixup on additional run). For neg_WT_6 we have too many reads and will need to only use one lane of data.  Before moving forward with any analyses we will first need to aggregate the counts across lanes for each unique sample, by taking the sum of counts for each individual gene.

```{r aggregate}

# Transpose data and add in sample info
tdata <- t(data)
tdata <- data.frame(tdata, samples=meta$samples)

# Aggregate on unique samples
unique_sample_mat <- aggregate(. ~ samples, data=tdata, FUN=sum)

# Transpose back and add in column names
data.combined <- t(unique_sample_mat)
colnames(data.combined) <- data.combined[1,]
data.combined <- data.matrix(data.combined[-1,]) 
class(data.combined) <- "numeric"

# Create a matching metadata file
meta.combined <- meta[,c('celltype', 'genotype', 'sort', 'samples')]
meta.combined <- unique(meta.combined)
row.names(meta.combined) <- meta.combined$samples
meta.combined <- meta.combined[colnames(data.combined),]

```

## Sequencing depth considering only exonic mapped reads
The number of reads for WT_sort6 is about 6X the number of reads for the other samples. This might be due to biases in the sample that change the composition of it resulting in 6x more prevalence in the pool. Over-amplification or insert size is way off?

In any case, we might need to remove this sample if it is biasing the results.


```{r genes-detected-plot}
dd = data.frame(Name=row.names(meta.combined), Seq.depth = colSums(data.combined))
ggplot(dd, aes(x=Name, y=Seq.depth)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90),
          axis.title = element_text(size = rel(1.5)),
        axis.text = element_text(size = rel(1.25))) +
    ylab("sequencing depth") + xlab("")
```


## Quality Control

### Sample-to-sample correlation matrix
To visualize clustering of data we will use transformed data by first running the count matrix through [DESeq2](https://www.bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.pdf). For certain comparison analysis using unsupervised techniques, it is useful to transform data. These techniques (i.e. PCA and clustering) perform better when values have a similar dynamic range. We use the rlog transform which is better for datasets in which there are large differences in sequencing depth.

An intercorrelation heatmap is plotted below generated by taking a correlation of transformed values for all pairwise combinations of samples. The samples show high correlations with each other (values higher than 0.99) indicating no major outliers. **The clustering pattern indicates two major groups associated with Neg and Tbr2+, and some separation of KO and WT within the Tbr2+ group.** 

```{r cluster-ica, fig.align='center', echo=FALSE, warning=FALSE, message=FALSE}

# Create DESeq2 dataset
meta.combined$group <- factor(paste(meta.combined$celltype, meta.combined$genotype, sep=":"))
dds <- DESeqDataSetFromMatrix(countData = data.combined, colData = meta.combined, design = ~group)
dds <- DESeq(dds)

# Matrix of pseudocounts for downstream visualization: two methods
rld <- rlog(dds, blind = TRUE)

# Heatmap of sample-to-sample correlation matrix
annotation <- data.frame(meta.combined[,c('celltype', 'genotype', 'sort')], row.names=row.names(meta.combined))
pheatmap(cor(assay(rld)), color = heatcolors.1, cluster_rows = T, annotation=annotation, border_color=NA)

```

### PCA of samples
Using all genes, we plotted a PCA plot of the first two principal components. The first principal component explains 66.2% of the variance and samples appear to segregate by celltype. This means the difference between cortical cells and committed neuron progenitor cells are large. For PC2, while a much smaller percentage of variance is explained, we do see that for this component samples separate to some extent based on WT and KO. Since there is some effect observed with the combined data, when separating samples by celltype we will see a more obvious segregation.


```{r pca, warning=FALSE, fig.align='center', echo=FALSE, fig.width=15}

# Perform PCA (requires transformed data matrix)
pca <- prcomp(t(assay(rld))

# Create data frame for input to ggplot
df <- cbind(meta.combined, pca$x[,c('PC1', 'PC2')])

# Plot with sample names used as data points
ggplot(df, aes(PC1, PC2, color =group)) + 
  theme_bw() +
  geom_point(size=4) +
  xlab('PC1 (66.2% variance explained)') +
  ylab('PC2 (0.05% variance explained)') +
  scale_x_continuous(expand = c(0.3,  0.3)) +
  theme(plot.title = element_text(size = rel(1.5)),
        axis.title = element_text(size = rel(1.5)),
        axis.text = element_text(size = rel(1.25)))

```

## Differential Expression in Tbr2+ cells
Starting from count data; we had created a DESeq object. The design formula (~) is used to estimate the dispersions and to estimate the log2 fold changes of the model and p-values based on how significantly genes are changing between WT and KO. A summary of the results are provided below. **At an adjusted p-value < 0.05, there are 418 genes up-regulated and 247 genes down-regulated.** *These expression changes are relative to the WT.*

```{r summary}

# Extract data of specified contrasts
# resultsNames(dds) # uncomment to check names for contrast setup
contrast <- list("groupTbr2.KO", "groupTbr2.WT")
resTbr <- results(dds, contrast=contrast)
summary(resTbr, alpha=p.cutoff)

```

